HDDS-13344. Fix ProxyUserAuthenticationFilter addition in HttpServer2

dombizita opened the pull request:
## What changes were proposed in this pull request?
Previously only the `AuthenticationFilterInitializer` was handled differently [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java#L648-L655). For AuthenticationFilterInitializer it'll call the `addFilter()` method which will be working for Ozone as well, as it'll use the config prefix key from Ozone. For the `initFilter()` it'll pick the config from the exact `AuthenticationFilterInitializer`, e.g. for `ProxyUserAuthenticationFilterInitializer` from [here](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/authentication/server/ProxyUserAuthenticationFilterInitializer.java#L39) it'll pick `hadoop.http.authentication.` (This was there as a [comment](https://github.com/apache/ozone/blame/d16354121d12405a1aeee2940f88512f989304ad/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java#L618C9-L618C9) in Ozone as well previous the latest change around there)

With this fix it'll handle `ProxyUserAuthenticationFilterInitializer` the correct way as well.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-13344

## How was this patch tested?

https://github.com/dombizita/ozone/actions/runs/15926689911
Tested it locally on a cluster

adoroszlai commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java:
```diff
@@ -649,6 +651,9 @@ private void initializeWebServer(Builder builder) throws IOException {
           addFilter("authentication",
               AuthenticationFilter.class.getName(), filterConfig);
+        } else if ((c instanceof ProxyUserAuthenticationFilterInitializer) && builder.securityEnabled) {
+          addFilter("authentication",
+              ProxyUserAuthenticationFilter.class.getName(), filterConfig);
```
`ProxyUserAuthenticationFilterInitializer` adds the filter with a different name, not `"authentication"`.  It also injects some extra settings on top of the config obtained from `AuthenticationFilter#getFilterConfigMap`.  See https://github.com/apache/hadoop/blob/626b227094027ed08883af97a0734d2db7863864/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/authentication/server/ProxyUserAuthenticationFilterInitializer.java#L42-L58

adoroszlai reviewed (commented):
Instead of trying to duplicate the logic from `AuthenticationFilterInitializer` and `ProxyUserAuthenticationFilter`, we could simply create a copy of the configuration and set the properties that these initializers expect.

```diff
diff --git hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
index 4b9f628223..1e0454a61b 100644
--- hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
+++ hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
@@ -643,15 +643,14 @@ private void initializeWebServer(Builder builder) throws IOException {
       builder.conf.set(BIND_ADDRESS, builder.hostName);
       org.apache.hadoop.conf.Configuration hadoopConf =
           LegacyHadoopConfigurationSource.asHadoopConfiguration(builder.conf);
-      Map<String, String> filterConfig = getFilterConfigMap(hadoopConf,
-          builder.authFilterConfigurationPrefix);
+      Map<String, String> filterConfig = getFilterConfigMap(hadoopConf, builder.authFilterConfigurationPrefix);
+      // create copy of the config with each <prefix>.<key> also added as hadoop.http.authentication.<key>
+      // (getFilterConfigMap removes prefix)
+      OzoneConfiguration copy = new OzoneConfiguration(hadoopConf);
+      filterConfig.forEach((k, v) -> copy.set("hadoop.http.authentication." + k, v));
+
       for (FilterInitializer c : initializers) {
-        if ((c instanceof AuthenticationFilterInitializer) && builder.securityEnabled) {
-          addFilter("authentication",
-              AuthenticationFilter.class.getName(), filterConfig);
-        } else {
-          c.initFilter(this, hadoopConf);
-        }
+        c.initFilter(this, copy);
       }
     }
 
```

What do you think?

dombizita commented:
> Instead of trying to duplicate the logic from `AuthenticationFilterInitializer` and `ProxyUserAuthenticationFilter`, we could simply create a copy of the configuration and set the properties that these initializers expect.
> 
> ```diff
> diff --git hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
> index 4b9f628223..a86d3c8cfb 100644
> --- hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
> +++ hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/HttpServer2.java
> @@ -645,13 +645,13 @@ private void initializeWebServer(Builder builder) throws IOException {
>            LegacyHadoopConfigurationSource.asHadoopConfiguration(builder.conf);
>        Map<String, String> filterConfig = getFilterConfigMap(hadoopConf,
>            builder.authFilterConfigurationPrefix);
> +      // create copy of the config with each <prefix>.<key> also added as hadoop.http.authentication.<key>
> +      // (getFilterConfigMap removes prefix)
> +      OzoneConfiguration copy = new OzoneConfiguration(hadoopConf);
> +      filterConfig.forEach((k, v) -> copy.set("hadoop.http.authentication." + k, v));
> +
>        for (FilterInitializer c : initializers) {
> -        if ((c instanceof AuthenticationFilterInitializer) && builder.securityEnabled) {
> -          addFilter("authentication",
> -              AuthenticationFilter.class.getName(), filterConfig);
> -        } else {
> -          c.initFilter(this, hadoopConf);
> -        }
> +        c.initFilter(this, copy);
>        }
>      }
>  
> ```
> 
> What do you think?

Thanks for the review @adoroszlai! Overall it's a similar way to do it, but while testing your suggestion I found that my implementation with the usage of `addFilter()` won't call the `getFilterConfigMap` method properly (with the impersonation configs in place), so it's better to go with your suggestion! I updated my patch accordingly.
